### ISSUE TITLE
Can not connect ingress to service

### ISSUE THREAD

**Issue by red-lichtie** [k3s-io/k3s#4123]
**Environmental Info:**
K3s Version:   v1.21.5+k3s1 (acad8ef8)
go version go1.16.8

uname -a
Node(s) CPU architecture, OS, and Version: Linux kube-primary 5.10.63-12-ARCH #1 SMP PREEMPT Fri Sep 24 05:19:28 UTC 2021 aarch64 GNU/Linux

Cluster Configuration:
Bare metal install on 4 x Raspberry Pi 4B

No Traefik  or klipper-lb (tried standard install but I was having the same error, actually it was worse because it didn't seem like any connections were getting to traefik as I described here when I tried talking to grafana: https://stackoverflow.com/questions/69351907/getting-prometheus-grafana-and-k3s-to-work-together)

**Describe the bug:**
I built myself a cluster to learn kubernetes and I read that k3s is the one of the easiest ways to get started on such small machines, so it is a new install on bare metal.

I installed as described here: https://www.puzzle.ch/de/blog/articles/2020/10/13/k3s-on-raspberry-pi

First I tried seeing the metrics-server with the following Ingress:  [metrics-ingress.yaml](https://github.com/k3s-io/k3s/files/7255361/metrics-ingress.txt)

Then I tried the simple "It works" example from the above page

[webserver-deployment.yaml](https://github.com/k3s-io/k3s/files/7255335/webserver-deployment.txt)
[webserver-ingress.yaml](https://github.com/k3s-io/k3s/files/7255336/webserver-ingress.txt)
[webserver-service.yaml](https://github.com/k3s-io/k3s/files/7255337/webserver-service.txt)

All I am getting is 504, regardless of what I did with the original traefik system or with the newly build one using metallb and nginx.

It looks like an IP address routing issue, some kind of security measure? Something I've missed in the Ingress definition(s)?

Connect to local-path-provisioner (10.42.0.3)
```
nslookup metrics-server.kube-system.svc.cluster.local
Server:         10.43.0.10
Address:        10.43.0.10:53

Name:   metrics-server.kube-system.svc.cluster.local
Address: 10.43.218.80
```

Connect to ingress-nginx-controller (10.42.3.4)
```
nslookup metrics-server.kube-system.svc.cluster.local
;; connection timed out; no servers could be reached

nslookup metrics-server.kube-system.svc.cluster.local 10.43.0.10
;; connection timed out; no servers could be reached
```

**Steps To Reproduce:**
- Installed K3s:
Controller:
```
export K3S_KUBECONFIG_MODE="644"
export INSTALL_K3S_EXEC=" --disable servicelb --disable traefik"
curl -sfL https://get.k3s.io | sh -
```

Worker nodes:
```
export K3S_KUBECONFIG_MODE="644"
export K3S_URL="https://192.168.76.200:6443"
export K3S_TOKEN="...."
curl -sfL https://get.k3s.io | sh -
```

I couldn't define any Ingress to begin with so I deleted the webhook (but that is an nginx problem).
```
kubectl delete -A ValidatingWebhookConfiguration ingress-nginx-admission
```

Controller:
```
helm repo add metallb https://metallb.github.io/metallb
helm install metallb stable/metallb --namespace kube-system     --set configInline.address-pools[0].name=default     --set configInline.address-pools[0].protocol=layer2     --set configInline.address-pools[0].addresses[0]=192.168.76.210-192.168.76.219
helm repo add ingress-nginx https://kubernetes.github.io/ingress-nginx
helm install ingress-nginx ingress-nginx/ingress-nginx --namespace kube-system --set defaultBackend.enabled=false
kubectl label node kube-worker1 node-role.kubernetes.io/worker=worker
kubectl label node kube-worker2 node-role.kubernetes.io/worker=worker
kubectl label node kube-worker3 node-role.kubernetes.io/worker=worker
```
I also had to disable some 

**Expected behavior:**
to see "It works" or the metrics.

**Actual behavior:**
504 Gateway Time-out

**Additional context / logs:**
nginx:
```
2021/09/29 22:17:58 [error] 728#728: *73412 upstream timed out (110: Operation timed out) while connecting to upstream, client: 10.42.3.1, server: _, request: "GET / HTTP/1.1", upstream: "http://10.42.1.6:80/", host: "192.168.76.210"
2021/09/29 22:18:03 [error] 728#728: *73412 upstream timed out (110: Operation timed out) while connecting to upstream, client: 10.42.3.1, server: _, request: "GET / HTTP/1.1", upstream: "http://10.42.1.6:80/", host: "192.168.76.210"
2021/09/29 22:18:08 [error] 728#728: *73412 upstream timed out (110: Operation timed out) while connecting to upstream, client: 10.42.3.1, server: _, request: "GET / HTTP/1.1", upstream: "http://10.42.1.6:80/", host: "192.168.76.210"
10.42.3.1 - - [29/Sep/2021:22:18:08 +0000] "GET / HTTP/1.1" 504 160 "-" "curl/7.79.1" 78 15.011 [webserver-httpd-80] [] 10.42.1.6:80, 10.42.1.6:80, 10.42.1.6:80 0, 0, 0 5.004, 5.004, 5.004 504, 504, 504 9961bee20685f33c9ddbba9877054a7f
```
Cluster
```kubectl get nodes -o wide
NAME           STATUS   ROLES                  AGE    VERSION        INTERNAL-IP      EXTERNAL-IP   OS-IMAGE         KERNEL-VERSION    CONTAINER-RUNTIME
kube-primary   Ready    control-plane,master   111m   v1.21.5+k3s1   192.168.76.200   <none>        Arch Linux ARM   5.10.63-12-ARCH   containerd://1.4.9-k3s1
kube-worker1   Ready    worker                 107m   v1.21.5+k3s1   192.168.76.201   <none>        Arch Linux ARM   5.10.63-12-ARCH   containerd://1.4.9-k3s1
kube-worker2   Ready    worker                 107m   v1.21.5+k3s1   192.168.76.202   <none>        Arch Linux ARM   5.10.63-12-ARCH   containerd://1.4.9-k3s1
kube-worker3   Ready    worker                 107m   v1.21.5+k3s1   192.168.76.203   <none>        Arch Linux ARM   5.10.63-12-ARCH   containerd://1.4.9-k3s1

kubectl get pods -o wide --all-namespaces 
NAMESPACE     NAME                                        READY   STATUS    RESTARTS   AGE    IP               NODE           NOMINATED NODE   READINESS GATES
kube-system   coredns-7448499f4d-6sfwb                    1/1     Running   0          109m   10.42.0.4        kube-primary   <none>           <none>
kube-system   metrics-server-86cbb8457f-wtg5g             1/1     Running   0          109m   10.42.0.2        kube-primary   <none>           <none>
kube-system   local-path-provisioner-5ff76fc89d-wcjtm     1/1     Running   0          109m   10.42.0.3        kube-primary   <none>           <none>
kube-system   metallb-controller-748756655f-4v9m6         1/1     Running   0          95m    10.42.2.2        kube-worker1   <none>           <none>
kube-system   metallb-speaker-8fptx                       1/1     Running   0          95m    192.168.76.200   kube-primary   <none>           <none>
kube-system   metallb-speaker-jb7kp                       1/1     Running   0          95m    192.168.76.201   kube-worker1   <none>           <none>
kube-system   metallb-speaker-wrvk5                       1/1     Running   0          95m    192.168.76.203   kube-worker3   <none>           <none>
kube-system   metallb-speaker-tq9cs                       1/1     Running   0          95m    192.168.76.202   kube-worker2   <none>           <none>
kube-system   ingress-nginx-controller-74cb6699df-vdplf   1/1     Running   0          77m    10.42.3.4        kube-worker2   <none>           <none>
webserver     httpd-554794f9fd-fw7c9                      1/1     Running   0          37m    10.42.1.6        kube-worker3   <none>           <none>

kubectl get svc -o wide --all-namespaces 
NAMESPACE     NAME                                 TYPE           CLUSTER-IP      EXTERNAL-IP      PORT(S)                      AGE    SELECTOR
default       kubernetes                           ClusterIP      10.43.0.1       <none>           443/TCP                      110m   <none>
kube-system   kube-dns                             ClusterIP      10.43.0.10      <none>           53/UDP,53/TCP,9153/TCP       110m   k8s-app=kube-dns
kube-system   metrics-server                       ClusterIP      10.43.218.80    <none>           443/TCP                      110m   k8s-app=metrics-server
kube-system   ingress-nginx-controller-admission   ClusterIP      10.43.23.66     <none>           443/TCP                      78m    app.kubernetes.io/component=controller,app.kubernetes.io/instance=ingress-nginx,app.kubernetes.io/name=ingress-nginx
kube-system   ingress-nginx-controller             LoadBalancer   10.43.236.80    192.168.76.210   80:30474/TCP,443:32150/TCP   78m    app.kubernetes.io/component=controller,app.kubernetes.io/instance=ingress-nginx,app.kubernetes.io/name=ingress-nginx
webserver     httpd                                ClusterIP      10.43.110.210   <none>           80/TCP                       42m    app=httpd

kubectl get ingress --all-namespaces 
NAMESPACE     NAME              CLASS    HOSTS   ADDRESS          PORTS   AGE
kube-system   metrics-ingress   <none>   *       192.168.76.210   80      64m
webserver     httpd-ingress     <none>   *       192.168.76.210   80      25m
```

**Backporting**
No idea

**Comment by brandond** [comment]
There's a lot there to unpack, but it looks like you're saying that DNS isn't working between pods running on different nodes. This is usually caused by some issue that's blocking CNI traffic. What Linux distro are you running on your Raspberry Pis? Is firewalld/ufw disabled?

Some users have reported issues with the flannel vxlan backend on recent kernels due to a bug in the kernel checksum offload code; you might try restarting with `--flannel-backend=host-gw`?

**Comment by red-lichtie** [comment]
As shown in 1st part of the Cluster information "Arch Linux ARM   5.10.63-12-ARCH", distro is Arch Linux ARM.

I haven't defined or installed any firewall rules or tools.

Is CNI maybe blocking the traffic? I saw that there are default "/16" networks on installation page.

I was assuming that the ingress pod could connect to all services, do I have to tell the networks that?

I am getting to the Nginx Pod, it can't get to the services though.

The installation only says "Uses the host-gw backend.", what that means isn't described so I left it to the default, and hopefully sane, option.

I'm trying to learn it so I don't know flannel or what host-gw would do, I'm more than happy to change it this evening when I get home though.

I'm also guessing now that the networking layer is preconfigured as secure as possible and that I should somehow know how to change the rules. Something about RBAC, Namespaces, something else?

**Comment by red-lichtie** [comment]
I added it to k3s start:
```
### Editing /etc/systemd/system/k3s.service.d/override.conf
### Anything between here and the comment below will become the new contents of the file

[Service]
ExecStart=
ExecStart=/usr/local/bin/k3s \
    server \
      '--disable' \
      'servicelb' \
      '--disable' \
      'traefik' \
      '--flannel-backend=host-gw'

### Lines below this comment will be discarded
```
Restarted the cluster and . . .
```
$ curl http://192.168.76.210:80/
<html><body><h1>It works!</h1></body></html>
```
That was the solution, I don't know what the difference is but who cares, as httpd says "It works!".

Thank you!

I think it would be nice if this was added somewhere to the k3s "Quick start" or "Installation" guide(s).

**Comment by brandond** [comment]
We haven't added it to the quickstart because vxlan is the best choice for most people. There's a kernel bug in the vxlan module that affected a limited number of fairly recent kernel versions, but some vendors haven't pulled it in yet.

**Comment by red-lichtie** [comment]
Hi Brad,

Thanks again for helping me get the cluster going, I was beginning to think I was too thick for this K8s thing. Now with cluster running, it is now back to my normal "just need to learn it" mode.

I'll flatten it again at the weekend and try it out with the default s (klipper-lb and traefik), that's what a lab cluster is for, right? :-)

Regarding the kernel patch, my Arch distro is currently on Kernel version 5.10.63, so I'm curious as to what version this kernel bug was patched (or released) in?

Could it already be patched and there might be a problem with the ARM architecture?

**Comment by brandond** [comment]
https://github.com/torvalds/linux/commit/ea64d8d6c675c0bb712689b13810301de9d8f77a

**Comment by red-lichtie** [comment]
Thanks, that commit made it into the kernel here: [https://cdn.kernel.org/pub/linux/kernel/v5.x/ChangeLog-5.6.13](https://cdn.kernel.org/pub/linux/kernel/v5.x/ChangeLog-5.6.13), so my problem is more likely to be "vxlan" and it's configuration (namespaces, etc.).

**Comment by brandond** [comment]
I'm not sure if it's also fixed on the 5.10 branch as well?
